File: 0049_group_anagrams/group_anagrams.cpp
#include <iostream>
#include <algorithm>
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

/**
 * Time: O(m * n * n * log(n))
 * Space: O(m * n * n)
 */
vector<vector<string>> group_anagrams(vector<string> &strs)
{
    unordered_map<string, vector<string>> map;
    vector<vector<string>> res;
    for (auto str : strs)
    {
        string copy = str;
        sort(copy.begin(), copy.end());
        map[copy].push_back(str);
    }
    for (auto m : map)
        res.push_back(m.second);
    return res;
}

File: 0049_group_anagrams/group_anagrams.js
/**
 * Time:    O(m * n * n * log(n))
 * Space:   O(m * n * n)
 * @param   {string[]}    strs
 * @return  {string[][]}
 */
const groupAnagrams = strs => {
  const map = {}
  for (let s of strs) {
    const sorted = s
      .split("")
      .sort()
      .join("")
    if (!(sorted in map)) map[sorted] = [s]
    else map[sorted].push(s)
  }
  return Object.values(map)
}
